# Lecture 13:  Symbolic Computation using SymPy

Estimated time needed: **50** minutes

### What this lecture is about
We will discuss the symbolic mathematics we have been using in school and colleges.

* Symbols
* Algebra
* trigsimp
* Calculas


This Notebook is created using Python

*   In case of any doubts contact to webinars@theanalytix.in 


## What is SymPy ? 
`SymPy` is a Python library for symbolic mathematics.

## What is Symbolic Computation?
`Symbolic computation` deals with the computation of mathematical objects symbolically. This means that the mathematical objects are represented exactly, not approximately, and mathematical expressions with unevaluated variables are left in symbolic form.

## Importing Package and check its Version

In [2]:
import sympy as sym

In [3]:
sym.__version__

'1.10.1'

## What are Symbols ? 
To define variables, we must use `symbols`. symbols takes a string of variable names separated by spaces or commas, and creates Symbols out of them. We can then assign these to variable names.

__Note__:  we note that the name of a Symbol and the name of the variable it is assigned to need not have anything to do with one another.



In [4]:
x = sym.Symbol('a')
x

a

multiple assignments will be done with `symbols`

In [5]:
a,b,c = sym.symbols('a b c')

In [6]:
a,b,c

(a, b, c)

In [7]:
b

b

In [16]:
import sympy
print(dir(sympy))

['Abs', 'AccumBounds', 'Add', 'Adjoint', 'AlgebraicField', 'AlgebraicNumber', 'And', 'AppliedPredicate', 'Array', 'AssumptionsContext', 'Atom', 'AtomicExpr', 'BasePolynomialError', 'Basic', 'BlockDiagMatrix', 'BlockMatrix', 'CC', 'CRootOf', 'Catalan', 'Chi', 'Ci', 'Circle', 'CoercionFailed', 'Complement', 'ComplexField', 'ComplexRegion', 'ComplexRootOf', 'Complexes', 'ComputationFailed', 'ConditionSet', 'Contains', 'CosineTransform', 'Curve', 'DeferredVector', 'DenseNDimArray', 'Derivative', 'Determinant', 'DiagMatrix', 'DiagonalMatrix', 'DiagonalOf', 'Dict', 'DiracDelta', 'DisjointUnion', 'Domain', 'DomainError', 'DotProduct', 'Dummy', 'E', 'E1', 'EPath', 'EX', 'EXRAW', 'Ei', 'Eijk', 'Ellipse', 'EmptySequence', 'EmptySet', 'Eq', 'Equality', 'Equivalent', 'EulerGamma', 'EvaluationFailed', 'ExactQuotientFailed', 'Expr', 'ExpressionDomain', 'ExtraneousFactors', 'FF', 'FF_gmpy', 'FF_python', 'FU', 'FallingFactorial', 'FiniteField', 'FiniteSet', 'FlagError', 'Float', 'FourierTransform', 'F

__Instead of writing sym for every time we can import all the functions of sympy using * operator__

In [15]:
from sympy import *
from sympy.abc import *

## Substitution
One of the most common things you might want to do with a mathematical expression is substitution. Substitution replaces all instances of something in an expression with something else. It is done using the `subs` method. 

In [19]:
expr = x**3 + 2*x**2 + 4*x + y
expr

x**3 + 2*x**2 + 4*x + y

In [23]:
expr.subs(x, -2)

y - 8

## Converting Strings to SymPy Expressions
The `sympify` function (that’s sympify, not to be confused with simplify) can be used to convert strings into SymPy expressions.

In [24]:
str_expr = "x**2 + 3*x - 1/2"
expr = sympify(str_expr)
expr

x**2 + 3*x - 1/2

In [25]:
expr.subs(x, 2)

19/2

## evalf
To evaluate a numerical expression into a floating point number, use `evalf`.

In [34]:
expr = sqrt(15)
expr1 = root(15, 2)
round(expr.evalf(),5), round(expr1.evalf(),5)

(3.87298, 3.87298)

## simplify 
SymPy has dozens of functions to perform various kinds of simplification. There is also one general function called simplify() that attempts to apply all of these functions in an intelligent way to arrive at the simplest form of an expression.

In [35]:
expr = sin(x)**2 + cos(x)**2
expr

sin(x)**2 + cos(x)**2

In [37]:
simplify(expr)

1

## expand
`expand()` is one of the most common simplification functions in SymPy. Although it has a lot of scopes, for now, we will consider its function in expanding polynomial expressions.

In [76]:
expr = (x-y)**2
expand(expr)

x**2 - 2*x*y + y**2

## factor
`factor()` takes a polynomial and factors it into irreducible factors over the rational numbers.

In [32]:
expr = (x**2*z + 4*x*y*z + 4*y**2*z)
expr

x**2*z + 4*x*y*z + 4*y**2*z

In [35]:
factor(expr)

z*(x + 2*y)**2

## collect
`collect()` collects common powers of a term in an expression.

In [78]:
expr = x*y + x - 3 + 2*x**2 - z*x**2 + x**3
expr

x**3 - x**2*z + 2*x**2 + x*y + x - 3

In [82]:
collect(expr, x)

x**3 + x**2*(2 - z) + x*(y + 1) - 3

## cancel

`cancel` takes input of rational function and put it down to simple mathematics cancelation.

In [85]:
expr = (x**2 - y**2) / (x+y)
expr

(x**2 - y**2)/(x + y)

In [42]:
cancel(expr)

x - y

In [44]:
expr = 1/x + (3*x/2 - 2)/(x-4)
expr

(3*x/2 - 2)/(x - 4) + 1/x

In [45]:
cancel(expr)

(3*x**2 - 2*x - 8)/(2*x**2 - 8*x)

## trigsimp

`trigsimp` reduces expression by using known trig identities<br>
Note: works on hyperbolic trig functions

In [46]:
expr = cosh(x)**2 + sinh(x)**2
expr

sinh(x)**2 + cosh(x)**2

In [47]:
trigsimp(expr)

cosh(2*x)

In [88]:
expr = sin(2*theta)
expr

sin(2*theta)

## expand_trig

In [90]:
expr = tan(A+B)
expr

tan(A + B)

In [91]:
expand_trig(expr)

(tan(A) + tan(B))/(-tan(A)*tan(B) + 1)

In [92]:
expand_trig((sin(alpha+beta)))

sin(alpha)*cos(beta) + sin(beta)*cos(alpha)

## Eq

In [52]:
Eq(expr, expand_trig(expr))

Eq(sin(2*theta), 2*sin(theta)*cos(theta))

In [40]:
Eq(x+y, 5)

Eq(x + y, 5)

## powsimp

In [94]:
expr = x**2 * x**3
expr

x**5

In [95]:
expr = x**3 * x**a
expr

x**3*x**a

In [55]:
powsimp(expr)

x**(a + 3)

In [60]:
expand_power_exp(y**(x-y))

y**x*y**(-y)

## log 

In [61]:
ln(x)

log(x)

In [62]:
log(y)

log(y)

## factorial

$0! = 1$<br>
$0! + 0! + 0! + 0! + 0! = (1 + 1 + 1 + 1 + 1) ! = 5! = 120$

In [65]:
factorial(factorial(0)+factorial(0)+factorial(0)+factorial(0)+factorial(0))

120

In [45]:
Eq(factorial(x), factorial(x), evaluate= False)

Eq(factorial(x), factorial(x))

## rational

In [72]:
Rational(1,2)

1/2

In [43]:
sympify('p/q')

p/q

In [97]:
Eq(sympify('7/3 + 3/4', evaluate= False), sympify('7/3 + 3/4'), evaluate = False)

Eq(3/4 + 7/3, 37/12)

In [48]:
a  = tan(theta)/ (1 - cot(theta))

In [49]:
b = cot(theta) / (1 - tan(theta))

In [50]:
a + b

tan(theta)/(1 - cot(theta)) + cot(theta)/(1 - tan(theta))

In [92]:
Eq(a+b, 1 + tan(theta)+ cot(theta))

Eq(tan(theta)/(1 - cot(theta)) + cot(theta)/(1 - tan(theta)), tan(theta) + cot(theta) + 1)

In [94]:
Eq(trigsimp(a+b), 1 + tan(theta)+ cot(theta))

Eq(tan(theta) + 1 + 1/tan(theta), tan(theta) + cot(theta) + 1)

In [106]:
expr = (1/sin(theta)**2 + 1/cos(theta)**2)
expr

cos(theta)**(-2) + sin(theta)**(-2)

In [111]:
factor(expr)

(sin(theta)**2 + cos(theta)**2)/(sin(theta)**2*cos(theta)**2)

In [105]:
(trigsimp(expr))

2*sqrt(2)*sqrt(-1/(cos(4*theta) - 1))

## Calculus

In [116]:
root(root(x,5),3)

x**(1/15)

In [100]:
expr = (x**2 + 4*x + 6)**5
expr

(x**2 + 4*x + 6)**5

### Diffrential

In [106]:
Eq(Derivative(expr,x), factor(diff(expr, x)))

Eq(Derivative((x**2 + 4*x + 6)**5, x), 10*(x + 2)*(x**2 + 4*x + 6)**4)

In [54]:
Eq(Derivative(tan(3*x),x), factor(diff(tan(3*x),x)))

Eq(Derivative(tan(3*x), x), 3*(tan(3*x)**2 + 1))

## Interagation

In [55]:
expr= sin(x)**3
expr

sin(x)**3

In [56]:
Integral(expr, x)

Integral(sin(x)**3, x)

In [57]:
Eq(Integral(expr, x), trigsimp(integrate(expr, x)))

Eq(Integral(sin(x)**3, x), cos(x)**3/3 - cos(x))

In [58]:
expr = -3*cos(x)/ 4
simplify(expr + cos(3*x)/ 12)

-3*cos(x)/4 + cos(3*x)/12

In [139]:
Integral(exp(-x))

Integral(exp(-x), x)

In [59]:
Eq(Integral(exp(-x), (x, 0, oo)), integrate(exp(-x), (x, 0, oo)))

Eq(Integral(exp(-x), (x, 0, oo)), 1)

In [60]:
ex = sin(x)**2 
ex

sin(x)**2

In [61]:
integrate(ex)

x/2 - sin(x)*cos(x)/2

## Limits

In [109]:
Limit(1/x, x, 0,'+')

Limit(1/x, x, 0)

In [67]:
Eq(Limit(1/y, y, 0), limit(1/y, y, 0))

Eq(Limit(1/y, y, 0), oo)

### solve

In [71]:
expr = q**2 + w*x + c
expr

c + q**2 + w*x

In [85]:
expr = Eq((1 / (x - 3)) - (1 / (x + 5)),Rational(1,6))
expr

Eq(-1/(x + 5) + 1/(x - 3), 1/6)

In [86]:
factor(expr)

Eq(8/((x - 3)*(x + 5)), 1/6)

In [87]:
solve(factor(expr))

[-9, 7]

In [89]:
expr = Eq(35*r**2  + (84- 65)*r - 156, 0)
expr

Eq(35*r**2 + 19*r - 156, 0)

In [90]:
solve(expr)

[-12/5, 13/7]

In [144]:
p1 = sympify("4", evaluate=False)
p2 = sympify("0.5 * 10**2", evaluate = False) / p1 
p3 = root(sympify('243 + 382', evaluate=False),2)
sympify("4 + Rational(1,2)*(10**2) / root(243 + 382, 2) * 2**10", evaluate=True)/ sympify('((60/100) * (400/8))/10', evaluate=False)

684

In [137]:
sympify("root(sympify('243 + 382', evaluate=False),2)* sympify('2**10', evaluate=False)")

1024*sqrt(243 + 382)


# Great Job!

That's all we need to know for now! Congratulations, you have completed one more lesson and hands-on with Python.
___
____

Proprietary Content. All Rights Reserved. This Notebook is prepared by <a href = 'https://www.linkedin.com/in/sumitkumarshukla'> Sumit Kumar Shukla</a> for SymPy Practice __© Copyright by The Analytix(Theax)__. Unauthorized use or distribution prohibited. All Questions Given Below is just made for enhancing your capabilities in Symbolic Mathematics.


## Logs


| Date (YYYY-MM-DD) | Version | Changed By | Change Description    |
| ----------------- | ------- | ---------- | --------------------- |
| 2022-11-01        | 1.1  | Purvi   | Prepared for Uploading |
| 2021-10-14        | 1.0    | Sumit   | Created Notebook Format of Lab  |
|                   |         |            |                       |
